### ISSUE TITLE
[no-unsafe-assignment] Optional params for Class constructor isn't any type

### ISSUE THREAD

**Issue by MathiasKandelborg** [typescript-eslint/typescript-eslint#2396]
I'm encountering a weird linting issue. It seems that Typescript is completely fine with, and understands that optional constructor parameters are their type if provided. This does not seem to be the case with ESLint.

**Rules**

```js
{
  "rules": {
        '@typescript-eslint/naming-convention': [
          'warn',
          {
            selector: 'variable',
            format: ['camelCase', 'PascalCase', 'UPPER_CASE']
          },
          {
            selector: 'interface',
            format: ['PascalCase'],
            prefix: ['I', 'A']
          },
          {
            selector: 'typeParameter',
            format: ['PascalCase'],
            prefix: ['T']
          },
          {
            selector: ['variable'],
            format: ['camelCase'],
            leadingUnderscore: 'allow'
          }
        ],
        'react/react-in-jsx-scope': 0,
        'react/prop-types': 0,
        'newline-before-return': 2,
        'import/no-unresolved': [1, { ignore: ['^(all|part):'] }],
        'comma-dangle': [
          'error',
          {
            arrays: 'never',
            objects: 'never',
            imports: 'never',
            exports: 'never',
            functions: 'never'
          }
        ]
  }
}
```

**Reproduction** can be done by:
1. Clone [my project](https://github.com/MathiasKandelborg/TRUE)
2. `yarn add sitemap`
3. Go to `/pages/api/sitemap.ts`
4. Edit the `smStream` const initialization to provide a string:
4.1 Make a host const: `const host = req?.headers?.host || 'not found'`
4.2. Edit smStream const: ``` hostname: `https://${host}` ```

--- 
**Additional infomation**

With the following type (from the sitemap lib):

```typescript
export declare class SitemapStream extends Transform {
    hostname?: string;
    level: ErrorLevel;
    hasHeadOutput: boolean;
    xmlNS: NSArgs;
    xslUrl?: string;
    private smiStream;
    lastmodDateOnly: boolean;
    constructor(opts?: SitemapStreamOptions);
    _transform(item: SitemapItemLoose, encoding: string, callback: TransformCallback): void;
    _flush(cb: TransformCallback): void;
}
```

And by assigning a variable to a class with above type:
```ts
    const host = req?.headers?.host || 'not found'
    const smStream = new SitemapStream({
      hostname: `https://${host}`
    }
```

The result is eslint complaining about an unsafe assignment and calling of an `any` value.

I do not have any typescript errors and can see the resolved type perfectly fine. ESLint is complaining though and I can't really tell why.

**More additional Info**

I was in a hurry to write this. It's edited to provide a clearer issue.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.6.1` |
| `@typescript-eslint/parser`        | `3.9.0` |
| `TypeScript`                       | `3.9.7` |
| `ESLint`                           | `7.6.0` |
| `node`                             | `14.7.0` | (might be another version for dev/production mode, this is my local installation)
| `npm`                              | `6.14.7` |

**Comment by bradzacher** [comment]
Repro steps:
1) clone your project
2) open in vscode
3) `yarn install`
4) open `pages/api/sitemap.ts`
5) See lots of TS errors and ESLint errors

![image](https://user-images.githubusercontent.com/7462525/90422462-bde34000-e06f-11ea-9a74-c29d98a0cbb4.png)

![image](https://user-images.githubusercontent.com/7462525/90422508-cfc4e300-e06f-11ea-9deb-f11a36c9a97e.png)

In particular, this one would definitely be the one that's causing the issue:
![image](https://user-images.githubusercontent.com/7462525/90422542-e10def80-e06f-11ea-9aaf-fbf1423e3dce.png)

**Comment by MathiasKandelborg** [comment]
I'm sorry I didn't provide a proper reproduction case. I haven't added a commit with the sitemap package the package.json file.

I've updated the issue to reflect steps needed to reproduce

**Comment by MathiasKandelborg** [comment]
I'm all over the place today for some reason, must be the heat and not enough liquids :sweat_smile: 

I have adequately updated the issue to provide some more precise information.

In the file, there are other errors. That's because I just copied some code without editing it at all.

The focus is that _after_ installing the sitemap package and having the provided types from it, the type of `smStream` is not undefined, but ESLint is reporting it as such.

**Comment by bradzacher** [comment]
Followed repro:

1) Clone my project
2) `yarn add -W sitemap`
3) Go to `/pages/api/sitemap.ts`
4) Edit the `smStream` const initialization to provide a string:
    1) Make a `host` const: `const host = req?.headers?.host || 'not found'`
    2) Edit `smStream` const: `` hostname: `https://${host}` ``

There are no errors for the construction:
![image](https://user-images.githubusercontent.com/7462525/90433807-4a4a2e80-e081-11ea-9b93-bf61f7ca4acf.png)

**Comment by MathiasKandelborg** [comment]
Okay, that is weird, I see the error. I've just updated the repository so that the file is implemented correctly now.


I didn't have a screenshot program, and while getting one accidentally closed my editor. The error is gone now, so I can't show it. I'm pretty sure it has something to do with my having my editor open for long periods. I am using vscode too, and while it's a bit modified, it is working just fine - compared to months ago where I'd experience OS freezes and worse - that is, until this falsy reporting from my eslint extension.

I'm a bit sad I didn't get to capture the error on my screen, but it probably had to do with typescript-eslint not updating or having it's available types updated after a `yarn add`.

**Comment by bradzacher** [comment]
Likely yes, this would probably be due to us not being able to detect changes in the node_modules folder.

On mac you can use `cmd + shift + 4` to capture a screenshot to desktop (include `ctrl` to capture to clipboard instead)
On windows10 you can use `shift + win + s` to capture a screenshot to clipboard.

**Comment by MathiasKandelborg** [comment]
Thanks for the tips, good to know, actually! Didn't know about the Win shortcut, I've never searched for it. For the past couple of years, I've used tiling WMs.

Oh wait, I think you're serving this to me on a silver platter:

I use Arch, btw
